src/libraries/System.Private.CoreLib/src/System/IO/FileStream.cs

```diff
@@ -223,38 +158,16 @@ public FileStream(string path, FileMode mode, FileAccess access, FileShare share
             if ((access & FileAccess.Read) != 0 && mode == FileMode.Append)
                 throw new ArgumentException(SR.Argument_InvalidAppendMode, nameof(access));
 
-            string fullPath = Path.GetFullPath(path);
-
-            _path = fullPath;
-            _access = access;
-            _bufferLength = bufferSize;
-
-            if ((options & FileOptions.Asynchronous) != 0)
-                _useAsyncIO = true;
-
             if ((access & FileAccess.Write) == FileAccess.Write)
             {
                 SerializationInfo.ThrowIfDeserializationInProgress("AllowFileWrites", ref s_cachedSerializationSwitch);
             }
 
-            _fileHandle = OpenHandle(mode, share, options);
-
-            try
-            {
-                Init(mode, share, path);
-            }
-            catch
-            {
-                // If anything goes wrong while setting up the stream, make sure we deterministically dispose
-                // of the opened handle.
-                _fileHandle.Dispose();
-                _fileHandle = null!;
-                throw;
-            }
+            _strategy = WrapForDerivedType(FileStreamStrategyHelper.ChooseStrategy(this, path, mode, access, share, bufferSize, options));
         }
 
         [Obsolete("This property has been deprecated.  Please use FileStream's SafeFileHandle property instead.  https://go.microsoft.com/fwlink/?linkid=14202")]
-        public virtual IntPtr Handle => SafeFileHandle.DangerousGetHandle();
+        public virtual IntPtr Handle => _strategy.SafeFileHandle.DangerousGetHandle(); // TODO: is it OK that this method has some logic?
```

Comment:
After introducing the `Strategy`, all public methods just validate the input argument and delegate the logic to the actual strategy. 

This property is an exception from this rule, as `SafeFileHandle.DangerousGetHandle()` is the actual logic. By leaving the TODO I wanted to get an opinion from the reviewers whether this is OK or not. What do you think?

Reply:
Is the concern whether we should use `_strategy.SafeFileHandle` or `SafeFileHandle` directly? If that's the case I would prefer the former for consistency with the other public members.

Reply:
My interpretation of the TODO question is to decide if we instead want something like:
```cs
public virtual IntPtr Handle => _strategy.GetHandle();
```
And have the strategy itself determine what the logic would be:
```cs
internal abstract class FileStreamStrategy : Stream
{
  ...
  public override IntPtr GetHandle() => SafeFileHandle.DangerousGetHandle();
  ...
}
```

But please correct me if I'm wrong, @adamsitnik.

If it's what I described, then I think I would prefer to have _all_ logic moved to the strategy itself. It would be consistent with other properties: `IsClosed`, `CanSeek`, `CanRead`, `CanWrite`, `Name`, `IsAsync`, etc.

If it's what @Jozkee described - Then I think having `_strategy.SafeFileHandle` would made make sense than calling `SafeFileHandle` directly, since the logic is _kind of_ deferred to the strategy, but this would not be consistent with how other properties are working.

Reply:
> My interpretation of the TODO question

Your interpretation is correct.

>  I would prefer to have all logic moved to the strategy itself 

done